server/ladder_service/ladder_service.py

```diff
@@ -673,6 +702,26 @@ async def launch_match(
                 if player not in connected_players
             ])
 
+    def calculate_dynamic_tokens_per_map(self, M: float, tokens: list[int]) -> float:
+        sorted_tokens = sorted(tokens)
+        if (sorted_tokens.count(0) >= M):
+            return 1
+
+        result = 1; last = 0; index = 0
+        while (index < len(sorted_tokens)): 
+            (index, last) = next(((i, el) for i, el in enumerate(sorted_tokens) if el > last), (len(sorted_tokens) - 1, sorted_tokens[-1]))
+            index += 1
+            divider = index - M
+            if (divider <= 0): 
+                continue
+            
+            result = sum(sorted_tokens[:index]) / divider
+            upperLimit = sorted_tokens[index] if index < len(sorted_tokens) else float('inf')
+            if (result <= upperLimit):
+                return result
+                
+        raise Exception("Failed to calculate dynamic tokens per map: impossible vetoes setup")
```

Comment:
Similarly here we are throwing an exception based on user input so a user could craft a veto selection that could cause the majority of games to fail to start. It would be better to have some default value rather than throwing an exception.